ouroboros-consensus/docs/utxo-hd/report.tex

```diff
@@ -0,0 +1,741 @@
+    \Statex
+    \State $\mathbb{L}^{s}_{[V_{in}]} \gets \texttt{stowLedgerTables}(L^{s}_{{V_{in}}})$
+    \State $\mathbb{L}' \gets \texttt{APPLY}(B, \mathbb{L}^{s}_{[V_{in}]})$
+    \State $\texttt{Diffs}_{B} \gets \texttt{diff}(\mathbb{L}', \mathbb{L}^{s}_{[V_{in}]})$
+    \Statex{}
+    \State \Return $L'_{(\texttt{Diffs}_{s}~\diamond~\texttt{Diffs}_{B})}$
+    \EndProcedure
+ \end{algorithmic}
+\end{algorithm}
+
+where \texttt{queryDisk} means consulting the Backing store and $\texttt{Diffs}_{DB}(Z)$ are the current diffs in the LedgerDB up to the requested block $Z$.
+
+\section{Anti-diffs}
+
+\subsection{The FingerTree problem}
+
+\subsection{Algebra of differences}
+
+\subsection{It is a semigroupoid!}
+
+\section{The Mempool}
+
+\subsection{Definition and design}
+
+In the Cardano node, the mempool plays a central role when forging the chain. Citing from the report:
+\begin{quote}
+  Whenever a block producing node is the leader of a slot, it gets the chance to
+  mint a block. For the Cardano blockchain to be useful, the minted block in the
+  blockchain needs to contain \emph{transactions}. The mempool is where we buffer
+  transactions until we are able to mint a block containing those transactions.
+\end{quote}
+
+We can consider the mempool as a stateful ubiquitous container that is accessed
+from multiple places at the same time and should be able to provide a list of
+transactions which are valid on top of a given state. And if possible, it should
+try to add transactions on top of the best known ledger state so that in the
+event that we need to forge on top of it, we don't need to revalidate all the
+transactions (or at least most invalid transactions have already been
+discarded). The interface to the mempool is roughly as follows:
+
+\begin{lstlisting}
+data Mempool m blk idx = Mempool {
+   tryAddTxs      :: WhetherToIntervene
+                  -> [GenTx blk]
+                  -> m ( [MempoolAddTxResult blk]
+                       , [GenTx blk]
+                       )
+ , syncWithLedger :: m (MempoolSnapshot blk idx)
+ , getSnapshotFor :: SlotNo
+                  -> TickedLedgerState blk DiffMK
+                  -> MempoolChangelog blk
+                  -> m (MempoolSnapshot blk idx)
+ , ...
+}
+\end{lstlisting}
+
+At every point in time, the mempool is anchored on top of the best ledger state
+known by the ChainDB, and a background thread is in charge of making sure that
+whenever the tip of the chain changes, the mempool gets revalidated.
+
+The mempool holds a \emph{virtual ledger state} resulting from ticking the
+anchor ledger state to some preemptive slot number and then applying all the
+currently valid transactions on top of it. Now that UTxO-HD has redefined what a
+ledger state is, the mempool holds not only a virtual ledger state but also a
```

Comment:
Maybe also applicable to previous instances: What about introducing clear nomenclature to stress the difference between a `LedgerState` in Consensus (which changed) and a ledger state for Ledger (which did not change)?